File: 77. Combinations/combine.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
const combine = (n, k) => {
  const ans = [];
  const dfs = (s, path) => {
    if (path.length === k) {
      ans.push([...path]);
      return;
    }

    for (let i = s; i <= n; i++) {
      path.push(i);
      dfs(i + 1, path);
      path.pop();
    }
  }

  dfs(1, []);
  return ans;
};


const n = 4;
const k = 2;
console.log('combine: ', combine(4, 2))

File: 77. Combinations/combine.py
from typing import List

class Solution:
  def combine(self, n: int, k: int) -> List[List[int]]:
    ans = []
    
    def dfs(s, path):
      if len(path) == k:
        ans.append(path.copy())
        return
      
      for i in range(s, n + 1):
        path.append(i)
        dfs(i + 1, path)
        path.pop()
        
    dfs(1, [])
    return ans
  
  
n = 4
k = 2
solution = Solution()
print('combine: ', solution.combine(4, 2))
